File: cpp-alg/algorithm/DoublyLinkedList.cpp
#include <iostream>

struct Node {
  int value;
  Node *prev;
  Node *next;
  Node(int value) : value(value), prev(nullptr), next(nullptr) {}
};

class DoublyLinkedList {
private:
  Node *head;
  Node *tail;

public:
  DoublyLinkedList() : head(nullptr), tail(nullptr) {}

  // add to tail
  void addToTail(int value) {
    Node *newNode = new Node(value);
    if (head == nullptr) {
      head = tail = newNode;
    } else {
      newNode->prev = tail;
      tail->next = newNode;
      tail = newNode;
    }
  }

  // add to head
  void addToHead(int value) {
    Node *newNode = new Node(value);

    if (head == nullptr) {
      head = tail = newNode;
    } else {
      newNode->next = head;
      head->prev = newNode;
      head = newNode;
    }
  }

  void display() {
    Node *current = head;
    while (current != nullptr) {
      std::cout << current->value << " ";
      current = current->next;
    }
    std::cout << std::endl;
  }
};

int main() {
  DoublyLinkedList list;

  list.addToTail(1);
  list.addToTail(2);
  list.addToTail(3);

  list.addToHead(4);

  list.display();
}

File: python-alg/basic/doubly_linked_list.py
class Node:
    def __init__(self, value):
        self.value = value
        self.prev = None
        self.next = None


class DoublyLinkedList:
    def __init__(self):
        self.head = None
        self.tail = None

    def add_to_tail(self, value):
        new_node = Node(value)
        if self.head is None:
            self.head = self.tail = new_node
        else:
            new_node.prev = self.tail
            self.tail.next = new_node
            self.tail = new_node

    def add_to_head(self, value):
        new_node = Node(value)
        if self.head is None:
            self.head = self.tail = new_node
        else:
            new_node.next = self.head
            self.head.prev = new_node
            self.head = new_node

    def display(self):
        current = self.head
        while current is not None:
            print(current.value)
            current = current.next


my_list = DoublyLinkedList()
my_list.add_to_tail(3)
my_list.add_to_tail(2)
my_list.add_to_tail(1)
my_list.add_to_head(4)
my_list.display()
